File: java(pure)/src/Hamiltonian.java
import java.util.*;
import java.util.stream.Collectors;

public class Hamiltonian {

    private Map<String, List<String>> usa;
    private List<String> easternStates;

    private List<String> journey = new ArrayList<>();
    private Map<String, List<String>> graph = new HashMap<>();

    public static void main(String[] args) {
        for (int i=0; i<30; ++i) {
            Hamiltonian h = new Hamiltonian();
            long start = Calendar.getInstance().getTimeInMillis();
            h.findHamiltonian("wdc");
            long end = Calendar.getInstance().getTimeInMillis();
            System.out.println("\nDuration: " + (end - start) + "ms");
            }
    }

    public Hamiltonian() {
        initialiseStates();
        this.graph.putAll(usa);
        this.reduceGraph(easternStates);
        this.checkGraph();
    }

    private void reduceGraph(List<String> statesToVisit){
        Map<String, List<String>> reducedGraph = new HashMap<>();
        for(String state : this.graph.keySet()) {
            if (statesToVisit.indexOf(state) >= 0) {
                List<String> neighbours = this.graph.get(state).stream().filter(
                    s -> statesToVisit.indexOf(s) >= 0).collect(Collectors.toList());
                reducedGraph.put(state, neighbours);
            }
        }
    this.graph = reducedGraph;
    }

    private void checkGraph() {
        for (Map.Entry<String, List<String>> entry : this.graph.entrySet()) {
            for (String neighbour : entry.getValue()) {
                if (this.graph.get(neighbour) == null) {
                    System.out.println("No graph entry for " + neighbour);
                } else if (this.graph.get(neighbour).indexOf(entry.getKey()) < 0){
                    System.out.println(entry.getKey() + " -> " + neighbour + " exists, but not " + neighbour + " -> " + entry.getKey());
                }
            }
        }
    }

    private void findHamiltonian(String start){
        if (!this.findHamiltoniamRecursively(start)) {
            System.out.println("No hamiltonian path found.");
        }
    }

    private boolean findHamiltoniamRecursively(String current){
        journey.add(current);

        if(journey.size() == graph.size()){
            for(String state : journey){
                System.out.print(state + "->");
            }
            return true;
        } else {
            for (String neighbour : graph.get(current)) {
                if (journey.indexOf(neighbour) < 0) {
                    if (findHamiltoniamRecursively(neighbour)) {
                        return true;
                    }
                }
            }
        }

        journey.remove(journey.size()-1);
        return false;
    }


    private void initialiseStates() {
        usa = new HashMap<>();
        usa.put("wa", Arrays.asList("or", "id"));
        usa.put("or", Arrays.asList("wa", "id", "nv", "ca"));
        usa.put("ca", Arrays.asList("or", "nv", "az"));
        usa.put("id", Arrays.asList("wa", "or", "nv", "ut", "wy", "mt"));
        usa.put("nv", Arrays.asList("or", "ca", "az", "ut", "id"));
        usa.put("ut", Arrays.asList("id", "nv", "az", "co", "wy"));
        usa.put("az", Arrays.asList("ca", "nv", "ut", "nm"));
        usa.put("mt", Arrays.asList("id", "wy", "sd", "nd"));
        usa.put("wy", Arrays.asList("mt", "id", "ut", "co", "ne", "sd"));
        usa.put("co", Arrays.asList("wy", "ut", "nm", "ok", "ks", "ne"));
        usa.put("nm", Arrays.asList("co", "az", "tx", "ok"));
        usa.put("nd", Arrays.asList("mt", "sd", "mn"));
        usa.put("sd", Arrays.asList("nd", "mt", "wy", "ne", "ia", "mn"));
        usa.put("ne", Arrays.asList("sd", "wy", "co", "ks", "mo", "ia"));
        usa.put("ks", Arrays.asList("ne", "co", "ok", "mo"));
        usa.put("ok", Arrays.asList("ks", "co", "nm", "tx", "ar", "mo"));
        usa.put("tx", Arrays.asList("ok", "nm", "la", "ar"));
        usa.put("mn", Arrays.asList("nd", "sd", "ia", "wi"));
        usa.put("ia", Arrays.asList("mn", "sd", "ne", "mo", "il", "wi"));
        usa.put("mo", Arrays.asList("ia", "ne", "ks", "ok", "ar", "tn", "ky", "il"));
        usa.put("ar", Arrays.asList("mo", "ok", "tx", "la", "ms", "tn"));
        usa.put("la", Arrays.asList("ar", "tx", "ms"));
        usa.put("wi", Arrays.asList("mn", "ia", "il"));
        usa.put("il", Arrays.asList("wi", "ia", "mo", "ky", "in"));
        usa.put("tn", Arrays.asList("ky", "mo", "ar", "ms", "al", "ga", "nc", "va", "ky"));
        usa.put("ms", Arrays.asList("tn", "ar", "la", "al", "tn"));
        usa.put("mi", Arrays.asList("in", "oh"));
        usa.put("in", Arrays.asList("mi", "il", "ky", "oh"));
        usa.put("ky", Arrays.asList("oh", "in", "il", "mo", "tn", "va", "wv", "oh"));
        usa.put("al", Arrays.asList("tn", "ms", "fl", "ga"));
        usa.put("ga", Arrays.asList("nc", "tn", "al", "fl", "sc"));
        usa.put("oh", Arrays.asList("mi", "in", "ky", "wv", "pa"));
        usa.put("wv", Arrays.asList("pa", "oh", "ky", "va", "md"));
        usa.put("ny", Arrays.asList("pa", "nj", "ct", "ma", "vt"));
        usa.put("nj", Arrays.asList("ny", "pa", "de"));
        usa.put("pa", Arrays.asList("ny", "nj", "oh", "wv", "md", "de"));
        usa.put("va", Arrays.asList("md", "wv", "ky", "tn", "nc", "wdc"));
        usa.put("nc", Arrays.asList("va", "tn", "ga", "sc"));
        usa.put("sc", Arrays.asList("nc", "ga"));
        usa.put("fl", Arrays.asList("ga", "al"));
        usa.put("me", Arrays.asList("nh"));
        usa.put("nh", Arrays.asList("me", "vt", "ma"));
        usa.put("vt", Arrays.asList("nh", "ny", "ma"));
        usa.put("ma", Arrays.asList("nh", "vt", "ny", "ct", "ri"));
        usa.put("ct", Arrays.asList("ma", "ny", "ri"));
        usa.put("ri", Arrays.asList("ma", "ct"));
        usa.put("de", Arrays.asList("pa", "md", "nj"));
        usa.put("md", Arrays.asList("pa", "wv", "va", "de", "wdc"));
        usa.put("wdc", Arrays.asList("md", "va"));

        easternStates = new ArrayList<>();
        easternStates.addAll(Arrays.asList("mn", "ia", "mo", "ar", "la", "wi", "il", "tn", "ms", "mi", "in", "ky", "al",
                "ga", "oh", "wv", "ny", "nj", "pa", "va", "nc", "sc", "fl", "me", "nh", "vt", "ma", "ct", "ri",
                "de", "md", "wdc"));
    }

}

File: c#/hamiltonian/Hamiltonian.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;
using static System.Array;

namespace hamiltonian
{
    public class Hamiltonian
    {
        private Dictionary<string, List<string>> _usa;
        private Array _easternStates;
        private Stack<string> _journey;
        
        public Hamiltonian()
        {
            InitialiseStates();
            ReduceGraph();
        }

        private void ReduceGraph() {
            foreach (var (state, neighbours) in _usa) {
                if (((IList) _easternStates).Contains(state)) {
                    neighbours.RemoveAll(neighbour => !((IList) _easternStates).Contains(neighbour));
                } else {
                    _usa.Remove(state);
                }
            }
        }

        public void PrintGraph() {
            foreach (var (state, neighbours) in _usa) {
                Console.WriteLine($"{state}: {String.Join(", ", neighbours)}");
            }
        }

        public void FindHamiltonian(string start) {
            _journey = new Stack<string>(26);
            
            if (!FindHamiltonianRecursively(start)) {
                Console.WriteLine("No hamiltonian path found.");
            }
        }
        private bool FindHamiltonianRecursively(string current) {
            _journey.Push(current);

            if (_journey.Count == _usa.Count) {
                Console.WriteLine(String.Join("->", _journey.Reverse()));
                return true;
            } else {
                foreach (var neighbour in _usa[current]) {
                    if (_journey.Contains(neighbour)) continue;
                    if (FindHamiltonianRecursively((neighbour))) return true;
                }
            }

            _journey.Pop();
            return false;
        }
        
        private void InitialiseStates() {
            _usa = new Dictionary<string, List<string>> {
                {"wa", new List<string> {"or", "id"}},
                {"or", new List<string> {"wa", "id", "nv", "ca"}},
                {"ca", new List<string> {"or", "nv", "az"}},
                {"id", new List<string> {"wa", "or", "nv", "ut", "wy", "mt"}},
                {"nv", new List<string> {"or", "ca", "az", "ut", "id"}},
                {"ut", new List<string> {"id", "nv", "az", "co", "wy"}},
                {"az", new List<string> {"ca", "nv", "ut", "nm"}},
                {"mt", new List<string> {"id", "wy", "sd", "nd"}},
                {"wy", new List<string> {"mt", "id", "ut", "co", "ne", "sd"}},
                {"co", new List<string> {"wy", "ut", "nm", "ok", "ks", "ne"}},
                {"nm", new List<string> {"co", "az", "tx", "ok"}},
                {"nd", new List<string> {"mt", "sd", "mn"}},
                {"sd", new List<string> {"nd", "mt", "wy", "ne", "ia", "mn"}},
                {"ne", new List<string> {"sd", "wy", "co", "ks", "mo", "ia"}},
                {"ks", new List<string> {"ne", "co", "ok", "mo"}},
                {"ok", new List<string> {"ks", "co", "nm", "tx", "ar", "mo"}},
                {"tx", new List<string> {"ok", "nm", "la", "ar"}},
                {"mn", new List<string> {"nd", "sd", "ia", "wi"}},
                {"ia", new List<string> {"mn", "sd", "ne", "mo", "il", "wi"}},
                {"mo", new List<string> {"ia", "ne", "ks", "ok", "ar", "tn", "ky", "il"}},
                {"ar", new List<string> {"mo", "ok", "tx", "la", "ms", "tn"}},
                {"la", new List<string> {"ar", "tx", "ms"}},
                {"wi", new List<string> {"mn", "ia", "il"}},
                {"il", new List<string> {"wi", "ia", "mo", "ky", "in"}},
                {"tn", new List<string> {"ky", "mo", "ar", "ms", "al", "ga", "nc", "va", "ky"}},
                {"ms", new List<string> {"tn", "ar", "la", "al", "tn"}},
                {"mi", new List<string> {"in", "oh"}},
                {"in", new List<string> {"mi", "il", "ky", "oh"}},
                {"ky", new List<string> {"oh", "in", "il", "mo", "tn", "va", "wv", "oh"}},
                {"al", new List<string> {"tn", "ms", "fl", "ga"}},
                {"ga", new List<string> {"nc", "tn", "al", "fl", "sc"}},
                {"oh", new List<string> {"mi", "in", "ky", "wv", "pa"}},
                {"wv", new List<string> {"pa", "oh", "ky", "va", "md"}},
                {"ny", new List<string> {"pa", "nj", "ct", "ma", "vt"}},
                {"nj", new List<string> {"ny", "pa", "de"}},
                {"pa", new List<string> {"ny", "nj", "oh", "wv", "md", "de"}},
                {"va", new List<string> {"md", "wv", "ky", "tn", "nc", "wdc"}},
                {"nc", new List<string> {"va", "tn", "ga", "sc"}},
                {"sc", new List<string> {"nc", "ga"}},
                {"fl", new List<string> {"ga", "al"}},
                {"me", new List<string> {"nh"}},
                {"nh", new List<string> {"me", "vt", "ma"}},
                {"vt", new List<string> {"nh", "ny", "ma"}},
                {"ma", new List<string> {"nh", "vt", "ny", "ct", "ri"}},
                {"ct", new List<string> {"ma", "ny", "ri"}},
                {"ri", new List<string> {"ma", "ct"}},
                {"de", new List<string> {"pa", "md", "nj"}},
                {"md", new List<string> {"pa", "wv", "va", "de", "wdc"}},
                {"wdc", new List<string> {"md", "va"}}
            };
            
            _easternStates = new[] {
                "mn", "ia", "mo", "ar", "la", "wi", "il", "tn", "ms", "mi", "in", "ky", "al",
                "ga", "oh", "wv", "ny", "nj", "pa", "va", "nc", "sc", "fl", "me", "nh", "vt", "ma", "ct", "ri",
                "de", "md", "wdc"
            };

        }
    }
}